### ISSUE TITLE
Instrument ES6 generators for coverage

### ISSUE THREAD

**Issue by Raynos** [gotwarlost/istanbul#66]
When using istanbul on ES6 code it prints a bunch of errors

``` js
Transformation error; return original code
{ [Error: Line 7: Unexpected token *] index: 174, lineNumber: 7, column: 9 }
Transformation error; return original code
{ [Error: Line 11: Unexpected token *] index: 266, lineNumber: 11, column: 9 }
Transformation error; return original code
{ [Error: Line 11: Unexpected token *] index: 323, lineNumber: 11, column: 9 }
Transformation error; return original code
{ [Error: Line 10: Unexpected token *] index: 241, lineNumber: 10, column: 9 }
Transformation error; return original code
{ [Error: Line 9: Unexpected token *] index: 206, lineNumber: 9, column: 9 }
Transformation error; return original code
{ [Error: Line 8: Unexpected token *] index: 223, lineNumber: 8, column: 9 }
Transformation error; return original code
{ [Error: Line 12: Unexpected token *] index: 280, lineNumber: 12, column: 9 }
Transformation error; return original code
{ [Error: Line 8: Unexpected token *] index: 174, lineNumber: 8, column: 9 }
Transformation error; return original code
{ [Error: Line 9: Unexpected token *] index: 194, lineNumber: 9, column: 9 }
Transformation error; return original code
{ [Error: Line 11: Unexpected token *] index: 260, lineNumber: 11, column: 13 }
Transformation error; return original code
{ [Error: Line 12: Unexpected token *] index: 302, lineNumber: 12, column: 13 }
```

The result is that every piece of code I have that uses generators is not instrumented.

**Comment by ChiChou** [comment]
Hey dude I got a way to run tests on ES6 code. Since esprima has already support ECMAScript6 but not merged to branch, you can manually install an alternative branch to enable the feature. Use git://github.com/jquery/esprima.git#harmony to replace the one in `node_modules/istanbul/node_modules/esprima`. Just use following commands:

```shell
cd node_modules/istanbul
npm install git://github.com/jquery/esprima.git#harmony
```

Also you can write these command in package.json to automate the installation:

```json
  "scripts": {
    "test": "make test-all",
    "postinstall": "pushd node_modules/istanbul&&npm install git://github.com/jquery/esprima.git#harmony"
  },
```

**Comment by ColCh** [comment]
For now: add this dependency to devDependencies:
```json
"istanbul": "git://github.com/gotwarlost/istanbul.git#harmony",
```
this will install istanbul from harmony. Good for me :+1:

**Comment by simonzack** [comment]
There's also a [istanbul-harmony](https://www.npmjs.com/package/istanbul-harmony#readme) package. It's rather old, could there be a merge from master?

**Comment by kunalgolani** [comment]
harmony branch is now 152 commits behind master. alternatives?

**Comment by tnagapavan** [comment]
I tried configuring Istanbul in my ember-cli application. And End up in getting the error of the same kind. Below is the details of the error.
group-test file contains the test cases and applied istanbul on that file. I tried with different approches like downgrading the Istanbul
version,using the code from harmony branch both i end up in getting the same exception.
Any inputs in related to resolving this issue will be very much helpful and appreciated.
D:\codecoverage\istanbul_kafka_1\tests\unit>istanbul cover group-test.js
Transformation error; return original code
{ [Error: Line 1: Unexpected reserved word]
  index: 15,
  lineNumber: 1,
  column: 16,
  description: 'Unexpected reserved word' }
No coverage information was collected, exit without writing coverage information
D:\codecoverage\istanbul_kafka_1\tests\unit\group-test.js:1
(function (exports, require, module, __filename, __dirname) { import { test, m
                                                              ^^^^^^
SyntaxError: Unexpected reserved word
    at exports.runInThisContext (vm.js:73:16)
    at Module._compile (module.js:443:25)
    at Object.Module._extensions..js (module.js:478:10)
    at Object.Module._extensions.(anonymous function) [as .js] (C:\Users\pavan\AppData\Roaming\npm\node_modules\istanbul\lib\hook.js:107:37)
    at Module.load (module.js:355:32)
    at Function.Module._load (module.js:310:12)
    at Function.Module.runMain (module.js:501:10)
    at runFn (C:\Users\pavan\AppData\Roaming\npm\node_modules\istanbul\lib\command\common\run-with-cover.js:122:16)
    at C:\Users\pavan\AppData\Roaming\npm\node_modules\istanbul\lib\command\common\run-with-cover.js:248:17
    at C:\Users\pavan\AppData\Roaming\npm\node_modules\istanbul\lib\util\file-matcher.js:68:16**

**Comment by albertorestifo** [comment]
I just want to :+1: this. Generators are amazing and should be supported, and not on an outdated harmony branch

**Comment by wavded** [comment]
:+1: - also experiencing issue

**Comment by kinshuk** [comment]
arrow functions work with istanbul version 0.3.14, but 'import' and 'export' usage gives error: Unexpected reserved word because esprima version is still 2.1.x.  esprima 2.2 supports import and export.  Any plans for new istanbul version that leverages esprima 2.2?

**Comment by Murphydbuffalo** [comment]
+1 @kinshuk

**Comment by gotwarlost** [comment]
@kinshuk - could you try the latest version of istanbul that uses the latest stable esprima? Please close this issue if it is not a problem any more. The harmony branch is now defunct and mainline has everything the harmony branch has and more.

**Comment by kinshuk** [comment]
Still seems unhappy...

$ ./node_modules/.bin/istanbul instrument test.js 
/tmp/node_modules/istanbul/node_modules/esprima/esprima.js:5474
            throw e;
                  ^
Error: Line 3: Unexpected token
    at createError (/tmp/node_modules/istanbul/node_modules/esprima/esprima.js:2336:21)
    at unexpectedTokenError (/tmp/node_modules/istanbul/node_modules/esprima/esprima.js:2411:13)
    at tolerateUnexpectedToken (/tmp/node_modules/istanbul/node_modules/esprima/esprima.js:2420:21)
    at parseStatementListItem (/tmp/node_modules/istanbul/node_modules/esprima/esprima.js:3814:21)
    at parseFunctionSourceElements (/tmp/node_modules/istanbul/node_modules/esprima/esprima.js:4649:23)
    at parseFunctionExpression (/tmp/node_modules/istanbul/node_modules/esprima/esprima.js:4859:16)
    at parsePrimaryExpression (/tmp/node_modules/istanbul/node_modules/esprima/esprima.js:3167:24)
    at inheritCoverGrammar (/tmp/node_modules/istanbul/node_modules/esprima/esprima.js:2592:18)
    at parseLeftHandSideExpressionAllowCall (/tmp/node_modules/istanbul/node_modules/esprima/esprima.js:3292:20)
    at inheritCoverGrammar (/tmp/node_modules/istanbul/node_modules/esprima/esprima.js:2592:18)

$ cat test.js 
'use strict';

export default {
   bar: 22
};


Digging deeper, it seems esprima wants sourceType = module
3809:    function parseStatementListItem() {
3810:        if (lookahead.type === Token.Keyword) {
3811:            switch (lookahead.value) {
3812:            case 'export':
3813:                if (sourceType !== 'module') {
3814:                    tolerateUnexpectedToken(lookahead, Messages.IllegalExportDeclaration);


I added "sourceType: 'module'" in ESP.parse call in instrumenter.js, but esprima died with a different error...

**Comment by RobinQu** [comment]
I've just converted a es5 project to es6 syntax including `class`, `super`. 

The coverage can be done, but it was logging something I don't understand.....

```

{ range: [ 646, 651 ],
  loc: 
   SourceLocation {
     start: Position { line: 34, column: 6 },
     end: { line: 34, column: 11 } },
  type: 'Super' }
Unsupported node type:Super
Transformation error; return original code
[TypeError: this[type] is not a function]
{ range: [ 618, 623 ],
  loc: 
   SourceLocation {
     start: Position { line: 26, column: 10 },
     end: { line: 26, column: 15 } },
  type: 'Super' }
Unsupported node type:Super
Transformation error; return original code
[TypeError: this[type] is not a function]
{ range: [ 172, 177 ],
  loc: 
   SourceLocation {
     start: Position { line: 7, column: 4 },
     end: { line: 7, column: 9 } },
  type: 'Super' }
Unsupported node type:Super
Transformation error; return original code
[TypeError: this[type] is not a function]
{ range: [ 816, 821 ],
  loc: 
   SourceLocation {
     start: Position { line: 23, column: 4 },
     end: { line: 23, column: 9 } },
  type: 'Super' }
Unsupported node type:Super
{ range: [ 1376, 1381 ],
  loc: 
   SourceLocation {
     start: Position { line: 45, column: 6 },
     end: { line: 45, column: 11 } },
  type: 'Super' }
Unsupported node type:Super
Transformation error; return original code
[TypeError: this[type] is not a function]
{ range: [ 363, 368 ],
  loc: 
   SourceLocation {
     start: Position { line: 13, column: 4 },
     end: { line: 13, column: 9 } },
  type: 'Super' }
Unsupported node type:Super
Transformation error; return original code
[TypeError: this[type] is not a function]
{ range: [ 521, 526 ],
  loc: 
   SourceLocation {
     start: Position { line: 22, column: 4 },
     end: { line: 22, column: 9 } },
  type: 'Super' }
Unsupported node type:Super
{ range: [ 1635, 1640 ],
  loc: 
   SourceLocation {
     start: Position { line: 66, column: 11 },
     end: { line: 66, column: 16 } },
  type: 'Super' }
Unsupported node type:Super
{ range: [ 1771, 1776 ],
  loc: 
   SourceLocation {
     start: Position { line: 74, column: 11 },
     end: { line: 74, column: 16 } },
  type: 'Super' }
Unsupported node type:Super
Transformation error; return original code
[TypeError: this[type] is not a function]
{ range: [ 559, 564 ],
  loc: 
   SourceLocation {
     start: Position { line: 15, column: 4 },
     end: { line: 15, column: 9 } },
  type: 'Super' }
Unsupported node type:Super
Transformation error; return original code
[TypeError: this[type] is not a function]
{ range: [ 362, 367 ],
  loc: 
   SourceLocation {
     start: Position { line: 17, column: 4 },
     end: { line: 17, column: 9 } },
  type: 'Super' }
Unsupported node type:Super
{ range: [ 1870, 1875 ],
  loc: 
   SourceLocation {
     start: Position { line: 75, column: 13 },
     end: { line: 75, column: 18 } },
  type: 'Super' }
Unsupported node type:Super
{ range: [ 2035, 2040 ],
  loc: 
   SourceLocation {
     start: Position { line: 81, column: 11 },
     end: { line: 81, column: 16 } },
  type: 'Super' }
Unsupported node type:Super
{ range: [ 2088, 2093 ],
  loc: 
   SourceLocation {
     start: Position { line: 85, column: 4 },
     end: { line: 85, column: 9 } },
  type: 'Super' }
Unsupported node type:Super
Transformation error; return original code
[TypeError: this[type] is not a function]
{ range: [ 183, 188 ],
  loc: 
   SourceLocation {
     start: Position { line: 8, column: 4 },
     end: { line: 8, column: 9 } },
  type: 'Super' }
Unsupported node type:Super
Transformation error; return original code
[TypeError: this[type] is not a function]
```

**Comment by davidrapin** [comment]
Fore those coming by and looking for a fix, istanbul@1.0.0-alpha.2 fixed this for me: https://github.com/gotwarlost/istanbul/releases/tag/v1.0.0-alpha.2

**Comment by hong-boy** [comment]
@davidrapin Thx for your suggestion!
